addons/mail/static/src/messaging/entity/message/message.js

```diff
@@ -0,0 +1,443 @@
+         */
+        async toggleStar() {
+            await this.env.rpc({
+                model: 'mail.message',
+                method: 'toggle_message_starred',
+                args: [[this.id]]
+            });
+        }
+
+        //----------------------------------------------------------------------
+        // Private
+        //----------------------------------------------------------------------
+
+        /**
+         * @override
+         */
+        _createInstanceLocalId(data) {
+            return `${this.constructor.localId}_${data.id}`;
+        }
+
+        /**
+         * @override
+         */
+        _update(data) {
+            const {
+                attachment_ids,
+                author_id, author_id: [
+                    authorId,
+                    authorDisplayName
+                ] = [],
+                body = this.body || "",
+                channel_ids,
+                customer_email_data = this.customer_email_data,
+                customer_email_status = this.customer_email_status,
+                date,
+                email_from = this.email_from,
+                history_partner_ids,
+                id = this.id,
+                isTransient = this.isTransient || false,
+                is_discussion = this.is_discussion || false,
+                is_note = this.is_note || false,
+                is_notification = this.is_notification || false,
+                message_type = this.message_type,
+                model,
+                moderation_status = this.moderation_status,
+                module_icon = this.module_icon,
+                needaction_partner_ids,
+                record_name,
+                res_id,
+                snailmail_error = this.snailmail_error,
+                snailmail_status = this.snailmail_status,
+                starred_partner_ids,
+                subject = this.subject,
+                subtype_description = this.subtype_description,
+                subtype_id = this.subtype_id,
+                tracking_value_ids = this.tracking_value_ids || [],
+            } = data;
+
+            this._write({
+                body,
+                customer_email_data,
+                customer_email_status,
+                date: date
+                    ? moment(str_to_datetime(date))
+                    : this.date
+                        ? this.date
+                        : moment(),
+                email_from,
+                id,
+                isTransient,
+                is_discussion,
+                is_note,
+                is_notification,
+                message_type,
+                model: 'mail.message',
+                moderation_status,
+                module_icon,
+                needaction_partner_ids,
+                snailmail_error,
+                snailmail_status,
+                subject,
+                subtype_description,
+                subtype_id,
+                tracking_value_ids,
```

Comment:
Isn't there another syntax that doesn't require to copy/paste the same keys twice in the code?

Reply:
Agree it's not nice, but we can't do much better due to how flexible we want to make changes to entity attributes/relations based on data.
I think we must solve this problem by having an API to define attributes, a bit similar to how relations work (but more powerful than this, because we need compute, dependencies, default values, etc.).

Wouldn't change anything for now.